javascript/implementation/Role.ts

```diff
@@ -0,0 +1,124 @@
+     * @param key either a Muid or container to check if it is included
+     * @param asOf optional timestamp to look back to
+     * @returns a promise that resolves to a boolean stating whether the key is included
+     */
+    async contains(key: Muid|Container, asOf?: AsOf): Promise<boolean> {
+        if ("address" in key) {
+            key = key.address;
+        }
+        const entry = await this.ginkInstance.store.getEntryByKey(this.address, key, asOf);
+        return Boolean(entry);
+    }
+
+    /**
+     * Function to iterate over the contents of the role.
+     * @param asOf optional timestamp to look back to
+     * @returns an async iterator across everything in the role, with values returned as MuidTuples
+     */
+    get_member_ids(asOf?: AsOf): AsyncGenerator<MuidTuple|KeyType|[], void, unknown> {
+        const thisSet = this;
+        return (async function*(){
+            const entries = await thisSet.ginkInstance.store.getKeyedEntries(thisSet.address, asOf);
+            for (const [key, entry] of entries) {
+                yield entry.effectiveKey;
+            }
+        })();
+    }
+
+    /**
+     * Returns the content of the role as a set.
+     * @param asOf optional timestamp to look back to
+     * @returns a promise that resolves to a set of Muids
+     */
+    async toSet(asOf?: AsOf): Promise<Set<Muid>> {
+        const entries = await this.ginkInstance.store.getKeyedEntries(this.address, asOf);
+        const resultSet = new Set<Muid>();
+        for (const [key, entry] of entries) {
+            if (typeof (entry.effectiveKey) == "object" && !(entry.effectiveKey instanceof Uint8Array) && !(entry.effectiveKey instanceof Array)) {
+                resultSet.add(muidTupleToMuid(entry.effectiveKey));
+            }
+        }
+        return resultSet;
+    }
+
+    /**
+     * Generates a JSON representation of the data in the role.
+     * Mostly intended for demo/debug purposes.
+     * @param indent true to pretty print (not yet implemented)
+     * @param asOf optional timestamp to look back to
+     * @param seen (internal use only! This prevents cycles from breaking things)
+     * @returns a JSON string
+     */
+    async toJson(indent: number | boolean = false, asOf?: AsOf, seen?: Set<string>): Promise<string> {
+        //TODO(https://github.com/google/gink/issues/62): add indentation
+        ensure(indent === false, "indent not implemented");
+        if (seen === undefined) seen = new Set();
+        const mySig = muidToString(this.address);
+        if (seen.has(mySig)) return "null";
+        seen.add(mySig);
+        const asSet = await this.toSet(asOf);
+        let returning = "[";
+        let first = true;
+        for (const key of asSet) {
+            if (first) {
+                first = false;
+            }   else {
+                returning += ",";
+            }
+            returning += await toJson(muidToString(key), indent === false ? false : +indent + 1, asOf, seen);
```

Comment:
A role being a set of containers, it's serialization should be a list where each element of the list is the result of calling toJson on the thing it contains.  Although users may be aware that the address of each container is a muid, when they call toJson really what they're asking for is the data stored in the database, not the addresses of the data structures.